### ISSUE TITLE
Finetuning Issue

### ISSUE THREAD

**Issue by Shaheer66** [NVIDIA/NeMo#15479]
I am trying to finetune the Nvidia models on the Openslr and CORAA ASR with some youtube extracted audios but the WER is even more higher than the base mode.
Model added too much noise, the base model has an average accuracy on similar test set is 85% but on the finetuned model it decreases to 50%
I have tried with my code and the available finetuning code in the Nemo repo but it does not work
Below I am adding my finetuning code

```python
import os
import torch
import nemo.collections.asr as nemo_asr
from omegaconf import OmegaConf, open_dict
import lightning.pytorch as pl

# --- 1. SET PATHS (Based on your setup) ---
ROOT = os.getcwd()
TRAIN_MANIFEST = os.path.join(ROOT, "train_manifest_nemo.jsonl")
VAL_MANIFEST = os.path.join(ROOT, "val_manifest_nemo.jsonl")

# --- 2. LOAD MODEL ---
print("Loading model...")
model = nemo_asr.models.EncDecHybridRNNTCTCBPEModel.from_pretrained(
    model_name="nvidia/stt_pt_fastconformer_hybrid_large_pc"
)

# --- 3. THE ERROR FIX: Patch Missing Mandatory Values ---
# This block prevents the "Missing mandatory value: dir" and "manifest_filepath" errors
with open_dict(model.cfg):
    # Fix the tokenizer directory crash
    if 'tokenizer' in model.cfg:
        model.cfg.tokenizer.dir = "" 
    
    # Fill placeholders for all data sections to satisfy OmegaConf
    for ds in ['train_ds', 'validation_ds', 'test_ds']:
        if ds in model.cfg:
            model.cfg[ds].manifest_filepath = TRAIN_MANIFEST if ds == 'train_ds' else VAL_MANIFEST
            model.cfg[ds].batch_size = 1 # Keep it 1 for CPU stability
            model.cfg[ds].num_workers = 0 
            model.cfg[ds].pin_memory = False

# --- 4. SETUP DATA ---
print("Setting up data loaders...")
model.setup_training_data(model.cfg.train_ds)
model.setup_validation_data(model.cfg.validation_ds)

# --- 5. RESULT ORIENTED: Protect the 85% Accuracy Baseline ---
# We freeze the encoder so your small pilot data doesn't "break" the model.
# This fixes the "one-word output (ela)" issue.
model.encoder.freeze()
print("Encoder frozen. Only fine-tuning Decoders for the pilot.")

# --- 6. OPTIMIZATION SETUP ---
model.setup_optimization(
    optim_config={
        'lr': 1e-4,
        'weight_decay': 0.001,
        'sched': {
            'name': 'CosineAnnealing',
            'warmup_steps': 100,
            'min_lr': 1e-6,
        },
    }
)

# --- 7. TRAINER (Strictly CPU-Compatible) ---
trainer = pl.Trainer(
    max_epochs=5,
    accelerator="cpu", # Change to "gpu" once you are on the RTX 4080
    devices=1,
    precision=32,      # 16-bit is for GPU only; 32-bit is required for CPU
    enable_checkpointing=True,
    logger=False
)

# --- 8. EXECUTION ---
print("Starting Pilot Training...")
trainer.fit(model)

# --- 9. SAVE FINAL MODEL ---
model.save_to("pt_br_pilot_final.nemo")
print("Successfully saved: pt_br_pilot_final.nemo")
```

Any support from your side is highly appreciated, Its not a personal project, it is my job requirement from client side we are targeting the more tahn 90% of accuracy so if anyone can guide me help me
It will be a great favour and I am very thankful to you for this guidance

**Comment by jeremy110** [comment]
If you're interested in welcoming a reference to this [GitHub](https://github.com/jeremy110/Finetune_Nemo_ASR), it contains the YAML conf, as well as support for the 8-bit optimizer.

**Comment by pzelasko** [comment]
Can you try using this script directly, and only modify the YAML config or overwrite only the options you need: https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/speech_to_text_finetune.py ?

**Comment by Shaheer66** [comment]
@pzelasko  thanks for your consideration, my issue was resolved using the below script

import os
import nemo.collections.asr as nemo_asr
from omegaconf import open_dict
import lightning.pytorch as pl

from lightning.pytorch.loggers import TensorBoardLogger

log_dir = "/home/ds-02/nemo_code/training_logs"
os.makedirs(log_dir, exist_ok=True)

logger = TensorBoardLogger(save_dir=log_dir, name="asr_finetune")
#

#ROOT = os.getcwd()

# TRAIN_MANIFEST = os.path.join(ROOT, "train_manifest_nemo.jsonl")
# VAL_MANIFEST = os.path.join(ROOT, "val_manifest_nemo.jsonl")

TRAIN_MANIFEST = "/home/ds-02/nemo_code/preprocessed_data/train_manifest.json"
VAL_MANIFEST   = "/home/ds-02/nemo_code/preprocessed_data/val_manifest.json"


def main():

    print("Loading pretrained model...")

    model = nemo_asr.models.EncDecHybridRNNTCTCBPEModel.from_pretrained(
        model_name="nvidia/stt_pt_fastconformer_hybrid_large_pc"
    )


    with open_dict(model.cfg):

        # tokenizer fix (avoids NeMo crash)
        if "tokenizer" in model.cfg:
            model.cfg.tokenizer.dir = ""

        # dataset configs
        for ds in ["train_ds", "validation_ds", "test_ds"]:

            if ds in model.cfg:

                if ds == "train_ds":
                    model.cfg[ds].manifest_filepath = TRAIN_MANIFEST
                else:
                    model.cfg[ds].manifest_filepath = VAL_MANIFEST

                model.cfg[ds].batch_size = 2
                model.cfg[ds].num_workers = 0
                model.cfg[ds].pin_memory = False

   

    print("Preparing datasets...")

    model.setup_training_data(model.cfg.train_ds)
    model.setup_validation_data(model.cfg.validation_ds)


    print("Freezing encoder (decoder-only training)...")

    model.encoder.freeze()


    model.setup_optimization(
        optim_config={
            "lr": 5e-5,
            "weight_decay": 0.001,
            "sched": {
                "name": "CosineAnnealing",
                "warmup_steps": 50,
                "min_lr": 1e-6,
            },
        }
    )



    # trainer = pl.Trainer(

    #     accelerator="cpu",
    #     devices=1,
    #     precision=32,

    #     max_epochs=15,

    #     log_every_n_steps=5,
    #     enable_progress_bar=True,
    #     enable_checkpointing=True,

    #     num_sanity_val_steps=0
    # )
    trainer = pl.Trainer(
    accelerator="cpu",
    devices=1,
    precision=32,
    max_epochs=15,
    log_every_n_steps=5,
    enable_progress_bar=True,
    enable_checkpointing=True,
    num_sanity_val_steps=0,
    logger=logger
)

    print("Starting training...")

    trainer.fit(model)

    

    print("Saving model...")

    model.save_to("pt_br_cpu_finetuned.nemo")

    print("Training completed successfully")


 

if __name__ == "__main__":
    main()

One thing I experienced that accuracy can be pushed by testing different decoding strategies more than finetuning
